anywidget/_descriptor.py

```diff
@@ -0,0 +1,615 @@
+"""Descriptor for _repr_mimebundle_ attribute that manages the comm channel.
+
+`MimeBundleDescriptor()` takes the place of a `_repr_mimebundle_` method on a class.
+
+- `MimeBundleDescriptor` is a
+  [descriptor](https://docs.python.org/3/howto/descriptor.html). When the
+  `_repr_mimebundle_` attribute is accessed on an instance of the decorated class, a
+  `ReprMimeBundle` instance is created and returned.
+- A `ReprMimeBundle` is first and foremost a callable object that implements the
+  `_repr_mimebundle_` protocol that jupyter expects.  However, it also manages an
+  ipykernel Comm instance that is used to send the state of the python model to the
+  javascript view.  This is done lazily, so that the Comm is only created when the
+  `_repr_mimebundle_` is first accessed.
+- `ReprMimeBundle` has the logic to get/set the state of the python model, and will keep
+  the two in sync ("bind"/"unbind_instance" methods can be used to control this).
+"""
+
+from __future__ import annotations
+
+import contextlib
+import json
+import sys
+import warnings
+import weakref
+from dataclasses import asdict, is_dataclass
+from typing import TYPE_CHECKING, Any, Callable, Iterable, overload
+
+
+from ._util import put_buffers, remove_buffers
+from ._version import __version__
+from .widget import DEFAULT_ESM
+
+if TYPE_CHECKING:  # pragma: no cover
+    import traitlets
+    import pydantic
+    import psygnal
+
+    from ipykernel.comm import Comm
+    from typing_extensions import TypeGuard
+
+    from ._protocols import CommMessage
```

Comment:
Just so I'm understanding, the guarded import here lets us use `CommMessage` in type annotations below (without quotes) but `from __future__ import annotations` turns it into a string for the interpreter?

Reply:
correct.  the most complete (if "dryest") explanation is in the PEP: https://peps.python.org/pep-0563/)

also worth nothing, `_protocols` is actually never imported anywhere (yet) when running anywidget at runtime, since the only import is behind if TYPE_CHECKING.  but we'd need to add typing_extensions to the deps if it was imported outside of TYPE_CHECKING